File: backend/app/schemas/tag.py
from datetime import datetime

from pydantic import BaseModel


class TagBase(BaseModel):
    name: str
    parent_id: int | None = None
    description: str | None = None
    color: str | None = None


class TagCreate(TagBase):
    pass


class TagUpdate(BaseModel):
    name: str | None = None
    parent_id: int | None = None
    description: str | None = None
    color: str | None = None


class TagResponse(TagBase):
    id: int
    created_at: datetime

    model_config = {"from_attributes": True}


class AllocationTargetInfo(BaseModel):
    target_pct: float
    notes: str | None = None

    model_config = {"from_attributes": True}


class TagTree(TagResponse):
    investment_count: int = 0
    total_value: float = 0.0
    allocation_target: AllocationTargetInfo | None = None
    children: list["TagTree"] = []


class TagWithAllocation(TagResponse):
    target_pct: float | None = None
    children: list["TagWithAllocation"] = []


TagTree.model_rebuild()
TagWithAllocation.model_rebuild()

File: frontend/src/types/tag.ts
export interface Tag {
  id: number;
  name: string;
  parent_id: number | null;
  description: string | null;
  color: string | null;
  created_at: string;
}

export interface AllocationTargetInfo {
  target_pct: number;
  notes: string | null;
}

export interface TagTreeNode extends Tag {
  children: TagTreeNode[];
  allocation_target?: AllocationTargetInfo | null;
  investment_count?: number;
  total_value?: number;
}

export interface TagCreate {
  name: string;
  parent_id?: number | null;
  description?: string;
  color?: string;
}

export interface TagUpdate {
  name?: string;
  parent_id?: number | null;
  description?: string;
  color?: string;
}
